src/extensions/default/CodeFolding/main.js

```diff
@@ -0,0 +1,326 @@
+        if (editor) {
+            var cursor = editor.getCursorPos(), cm = editor._codeMirror;
+            cm.unfoldCode(cursor.line);
+        }
+    }
+
+    /**
+        Collapses all foldable regions in the current document. Folding is done up to a level 'n'
+        which is defined in the preferences. Levels refer to fold heirarchies e.g., for the following
+        code fragment, the function is level 1, the if statement is level 2 and the forEach is level 3
+
+            function sample() {
+                if (debug) {
+                    logMessages.forEach(function (m) {
+                        console.debug(m);
+                    });
+                }
+            }
+    */
+    function collapseAll() {
+        var editor = EditorManager.getFocusedEditor();
+        if (editor && editor._codeMirror) {
+            var cm = editor._codeMirror;
+            CodeMirror.commands.foldToLevel(cm);
+        }
+    }
+
+    /**
+        Expands all folded regions in the current document
+    */
+    function expandAll() {
+        var editor = EditorManager.getFocusedEditor();
+        if (editor && editor._codeMirror) {
+            var cm = editor._codeMirror;
+            CodeMirror.commands.unfoldAll(cm);
+        }
+    }
+
+    function createGutter(editor) {
+        var cm = editor._codeMirror;
+        if (cm) {
+            var path = editor.document.file.fullPath, _lineFolds = _prefs.get(path);
+            _lineFolds = _lineFolds || {};
+            cm._lineFolds = _lineFolds;
+            var gutters = cm.getOption("gutters").slice(0);
+            var lnIndex = gutters.indexOf("CodeMirror-linenumbers");
+            gutters.splice(lnIndex + 1, 0, gutterName);
+            cm.setOption("gutters",  gutters);
+            cm.setOption("foldGutter", {onGutterClick: onGutterClick});
+
+            $(cm.getGutterElement()).on({
+                mouseenter: function () {
+                    if (_prefs.getSetting("fadeFoldButtons")) {
+                        foldGutter.updateInViewport(cm);
+                    }
+                },
+                mouseleave: function () {
+                    if (_prefs.getSetting("fadeFoldButtons")) {
+                        foldGutter.clearGutter(cm);
+                    }
+                }
+            });
+        }
+    }
+
+    function onActiveEditorChanged(event, current, previous) {
+        if (_prefs.getSetting("enabled")) {
+            if (current && current._codeMirror.getOption("gutters").indexOf(gutterName) === -1) {
+                createGutter(current);
+                restoreLineFolds(current);
+            }
+            if (previous) {
+                saveLineFolds(previous);
+            }
+        }
+    }
+
+    function saveBeforeClose() {
+        saveLineFolds(EditorManager.getCurrentFullEditor());
+    }
+
+    /**
+        Initialise the extension
+    */
+    function init() {
+        if (!CodeMirror.fold.combine && _prefs.getSetting("enabled")) {
```

Comment:
In the name of readability, can we make this `if` statement be an early `return`?  This is merely to get rid of the deep indentation in the body of the `if` statement.